Add timeout thread safety

tznind opened the pull request:
Fixes  #1679

I think this should fix the above issue.  There are 2 problems addressed by this PR.  

There was locking on the `timeouts` field (i.e. `lock (timeouts)`).  But since this variable is reassigned to a new list in `RunTimers` the reference changes so locking will not be consistent.  I have fixed this by adding an unchanging object `timeoutsLockToken` and locking on that instead.

RunTimers iterates through a `copy` of `timeouts` but there was no protection of the period in which you take the copy but haven't yet assigned a new list for callers to AddTimeout to add new timeouts to which could result in various errors.  I've added an extra lock around that section.  Initially I put it around the whole `RunTimers` but I think that is excessive and the solution implemented works.

I've added a test which is quite conservative (100 threads x 100 timeout callbacks of 1 second x 5 seconds runtime) which shouldn't stress any machines running it much.

tig commented:
> I've added a test which is quite conservative (100 threads x 100 timeout callbacks of 1 second x 5 seconds runtime) which shouldn't stress any machines running it much.

Just so I'm clear, if this test runs without your fix, the bug is ~100% repro?

tznind commented:
> > I've added a test which is quite conservative (100 threads x 100 timeout callbacks of 1 second x 5 seconds runtime) which shouldn't stress any machines running it much.
> 
> Just so I'm clear, if this test runs without your fix, the bug is ~100% repro?

I have run against main and find that the test fails about 75% of the time, it is a race condition after all so hard to test sensibly.  


##Running against main##
```
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  Restored Z:\Repos\gui.cs\UICatalog\UICatalog.csproj (in 317 ms).
  Restored Z:\Repos\gui.cs\Example\Example.csproj (in 18 ms).
  Restored Z:\Repos\gui.cs\ReactiveExample\ReactiveExample.csproj (in 35 ms).
  Restored Z:\Repos\gui.cs\Terminal.Gui\Terminal.Gui.csproj (in 44 ms).
  Restored Z:\Repos\gui.cs\UnitTests\UnitTests.csproj (in 261 ms).
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:04.46]     Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [FAIL]
  Failed Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [575 ms]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at Terminal.Gui.MainLoop.RunTimers() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 180
   at Terminal.Gui.MainLoop.MainIteration() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 241
   at Terminal.Gui.Application.RunLoop(RunState state, Boolean wait) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 980
   at Terminal.Gui.Application.Run(Toplevel view, Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1127
   at Terminal.Gui.Application.Run(Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1069
   at Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts() in Z:\Repos\gui.cs\UnitTests\ApplicationTests.cs:line 1329

Failed!  - Failed:     1, Passed:   675, Skipped:     0, Total:   676, Duration: 7 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:07.63]     Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [FAIL]
  Failed Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [355 ms]
  Error Message:
   System.InvalidOperationException : Collection was modified after the enumerator was instantiated.
  Stack Trace:
     at System.Collections.Generic.SortedList`2.Enumerator.MoveNext()
   at Terminal.Gui.MainLoop.RunTimers() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 176
   at Terminal.Gui.MainLoop.MainIteration() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 241
   at Terminal.Gui.Application.RunLoop(RunState state, Boolean wait) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 980
   at Terminal.Gui.Application.Run(Toplevel view, Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1127
   at Terminal.Gui.Application.Run(Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1069
   at Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts() in Z:\Repos\gui.cs\UnitTests\ApplicationTests.cs:line 1329

Failed!  - Failed:     1, Passed:   675, Skipped:     0, Total:   676, Duration: 8 s - UnitTests.dll (net6.0)

PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:08.37]     Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [FAIL]
  Failed Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts [139 ms]
  Error Message:
   System.InvalidOperationException : Collection was modified after the enumerator was instantiated.
  Stack Trace:
     at System.Collections.Generic.SortedList`2.Enumerator.MoveNext()
   at Terminal.Gui.MainLoop.RunTimers() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 176
   at Terminal.Gui.MainLoop.MainIteration() in Z:\Repos\gui.cs\Terminal.Gui\Core\MainLoop.cs:line 241
   at Terminal.Gui.Application.RunLoop(RunState state, Boolean wait) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 980
   at Terminal.Gui.Application.Run(Toplevel view, Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1127
   at Terminal.Gui.Application.Run(Func`2 errorHandler) in Z:\Repos\gui.cs\Terminal.Gui\Core\Application.cs:line 1069
   at Terminal.Gui.Core.ApplicationTests.TestAddManyTimeouts() in Z:\Repos\gui.cs\UnitTests\ApplicationTests.cs:line 1329

Failed!  - Failed:     1, Passed:   675, Skipped:     0, Total:   676, Duration: 8 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   676, Skipped:     0, Total:   676, Duration: 13 s - UnitTests.dll (net6.0)
```

##Running against this branch##

```
PS Z:\Repos\gui.cs> git checkout add-timeout-thread-safety
Switched to branch 'add-timeout-thread-safety'
Your branch is up to date with 'origin/add-timeout-thread-safety'.
PS Z:\Repos\gui.cs> git pull
Already up to date.
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   676, Skipped:     0, Total:   676, Duration: 13 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   676, Skipped:     0, Total:   676, Duration: 13 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   676, Skipped:     0, Total:   676, Duration: 13 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net472\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\netstandard2.0\Terminal.Gui.dll
  Terminal.Gui -> Z:\Repos\gui.cs\Terminal.Gui\bin\Debug\net6.0\Terminal.Gui.dll
  UICatalog -> Z:\Repos\gui.cs\UICatalog\bin\Debug\net6.0\UICatalog.dll
  UnitTests -> Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll
Test run for Z:\Repos\gui.cs\UnitTests\bin\Debug\net6.0\UnitTests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   676, Skipped:     0, Total:   676, Duration: 13 s - UnitTests.dll (net6.0)
PS Z:\Repos\gui.cs>
```

tig reviewed (approved):
LGTM.
